### PR TITLE
Adding GetUsersAsync with filters and pagination

### PR DESCRIPTION
**Pull request by assyadh**
*Issue #, if available:* DOTNET-3255

*Description of changes:* Exposing the UserManager.Users property as a method (GetUsersAsync) with optional filters to filter the users.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/Amazon.AspNetCore.Identity.Cognito/CognitoUserManager.cs [hunk 1]
```diff
@@ -528,6 +528,49 @@ protected override Task<IdentityResult> UpdateUserAsync(TUser user)
...
+        ///     of Using the ListUsers API in the Amazon Cognito Developer Guide.</param>
+        /// <param name="attributeFilterType"> The type of filter to apply:
+        ///     For an exact match, use =
+        ///     For a prefix ("starts with") match, use ^=
+        /// </param>
+        /// <param name="attributeFilterValue"> The filter value for the specified attribute.</param>
+        /// <returns>
+        /// The <see cref="Task"/> that represents the asynchronous operation, containing a IEnumerable of CognitoUser.
+        /// </returns>
+        public virtual Task<IEnumerable<CognitoUser>> GetUsersAsync(string filterName = "", string filterType = "", string filterValue = "")
```
Comment (costleya) [inline src/Amazon.AspNetCore.Identity.Cognito/CognitoUserManager.cs]: FilterType should probably be an enum.

FilterName can follow the pattern of "Constant Classes" to have readable constants for each of the currently accepted filter keys, while also allowing users to enter their own if the API expands.

File: src/Amazon.AspNetCore.Identity.Cognito/CognitoUserManager.cs [hunk 2]
```diff
@@ -528,6 +528,49 @@ protected override Task<IdentityResult> UpdateUserAsync(TUser user)
             return _userStore.UpdateAsync(user, CancellationToken);
         }
 
+        /// <summary>
+        ///  Not supported: Returns an IQueryable of users if the store is an IQueryableUserStore
+        /// </summary>
+        public override IQueryable<TUser> Users
```
Comment (costleya) [inline src/Amazon.AspNetCore.Identity.Cognito/CognitoUserManager.cs]: _minor_
expression syntax would probably more readable and succinct.

``` csharp
public override IQueryable<TUser> Users => throw new NotSupportedException("This property is not supported. Use GetUsersAsync() instead.");
```

File: src/Amazon.AspNetCore.Identity.Cognito/CognitoUserStore/CognitoUserStore.cs [hunk 3]
```diff
@@ -227,6 +228,74 @@ public virtual async Task<IdentityResult> ResetPasswordAsync(TUser user, Cancell
...
+        ///     For an exact match, use =
+        ///     For a prefix ("starts with") match, use ^=
+        /// </param>
+        /// <param name="attributeFilterValue"> The filter value for the specified attribute.</param>
+        /// <returns>
+        /// The <see cref="Task"/> that represents the asynchronous operation, containing a IEnumerable of CognitoUser.
+        /// </returns>
+        public virtual async Task<IEnumerable<CognitoUser>> GetUsersAsync(string attributeFilterName, string attributeFilterType, string attributeFilterValue, CancellationToken cancellationToken)
+        {
+
```
Comment (costleya) [inline src/Amazon.AspNetCore.Identity.Cognito/CognitoUserStore/CognitoUserStore.cs]: _minor_
/w

File: src/Amazon.AspNetCore.Identity.Cognito/CognitoUserStore/CognitoUserStore.cs [hunk 4]
```diff
@@ -227,6 +228,74 @@ public virtual async Task<IdentityResult> ResetPasswordAsync(TUser user, Cancell
...
+                }
+
+                foreach (var user in response.Users)
+                {
+                    result.Add(new CognitoUser(user.Username, _pool.ClientID, _pool, _cognitoClient, null,
+                    user.UserStatus.Value, user.Username,
+                    user.Attributes.ToDictionary(attribute => attribute.Name, attribute => attribute.Value)));
+                }
+
+            } while (!string.IsNullOrEmpty(response.PaginationToken));
```
Comment (costleya) [inline src/Amazon.AspNetCore.Identity.Cognito/CognitoUserStore/CognitoUserStore.cs]: _remark_
this is why we need pager support in the SDK...

File: src/Amazon.AspNetCore.Identity.Cognito/CognitoAttribute.cs [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        public static readonly CognitoAttribute UserStatus = new CognitoAttribute("cognito:user_status");
+        public static readonly CognitoAttribute Website = new CognitoAttribute("website");
+        public static readonly CognitoAttribute ZoneInfo = new CognitoAttribute("zoneinfo");
+
+
+        // List of attributes filterable through the ListUsers API.
+        public static readonly string[] FilterableAttributes = { Email.AttributeName, Enabled.AttributeName, GivenName.AttributeName,
+            FamilyName.AttributeName, PhoneNumber.AttributeName, PreferredUsername.AttributeName, Name.AttributeName, Sub.AttributeName, UserName.AttributeName, UserStatus.AttributeName };
+
+        public string AttributeName { get; private set; }
```
Comment (costleya) [inline src/Amazon.AspNetCore.Identity.Cognito/CognitoAttribute.cs]: _minor_

The private readonly set is implicit

File: src/Amazon.AspNetCore.Identity.Cognito/CognitoAttribute.cs [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+        public CognitoAttribute(string attributeName)
```
Comment (costleya) [inline src/Amazon.AspNetCore.Identity.Cognito/CognitoAttribute.cs]: _suggestion_

public CognitoAttribute(string attributeName) => (AttributeName) = (attributeName);

File: src/Amazon.AspNetCore.Identity.Cognito/CognitoAttributeFilterType.cs [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+ */
+
+namespace Amazon.AspNetCore.Identity.Cognito
+{
+    public class CognitoAttributeFilterType
+    {
+        public static readonly CognitoAttributeFilterType IsEqualTo = new CognitoAttributeFilterType("=");
+        public static readonly CognitoAttributeFilterType StartsWith = new CognitoAttributeFilterType("^=");
+
+        private string FilterType { get; set; }
```
Comment (costleya) [inline src/Amazon.AspNetCore.Identity.Cognito/CognitoAttributeFilterType.cs]: private readonly set

File: src/Amazon.AspNetCore.Identity.Cognito/CognitoAttribute.cs (hunk 5 above)
Comment (assyadh) [reply src/Amazon.AspNetCore.Identity.Cognito/CognitoAttribute.cs]: The property needs to be public but the setter private.

File: src/Amazon.AspNetCore.Identity.Cognito/CognitoAttributeFilterType.cs (hunk 7 above)
Comment (assyadh) [reply src/Amazon.AspNetCore.Identity.Cognito/CognitoAttributeFilterType.cs]: Done

**Pull request merged**